sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/aggregate/datasketchesAggregates.scala

```diff
@@ -0,0 +1,523 @@
+              "inputName" -> "lgConfigK",
+              "inputType" -> toSQLType(second.dataType),
+              "inputExpr" -> toSQLExpr(second)
+            )
+          )
+        } else if (lgConfigK <= 0L) {
+          DataTypeMismatch(
+            errorSubClass = "VALUE_OUT_OF_RANGE",
+            messageParameters = Map(
+              "exprName" -> "lgConfigK",
+              "valueRange" -> s"[0, positive]",
+              "currentValue" -> toSQLValue(lgConfigK, IntegerType)
+            )
+          )
+        } else if (!third.foldable) {
+          DataTypeMismatch(
+            errorSubClass = "NON_FOLDABLE_INPUT",
+            messageParameters = Map(
+              "inputName" -> "numBitsExpression",
+              "inputType" -> toSQLType(third.dataType),
+              "inputExpr" -> toSQLExpr(third)
+            )
+          )
+        } else {
+          TypeCheckSuccess
+        }
+      case _ =>
+        DataTypeMismatch(
+          errorSubClass = "HLLSKETCH_WRONG_TYPE",
+          messageParameters = Map(
+            "functionName" -> toSQLId(prettyName),
+            "expectedSecond" -> toSQLType(IntegerType),
+            "expectedThird" -> toSQLType(StringType),
+            "actual" -> Seq(first.dataType, second.dataType, third.dataType)
+              .map(toSQLType).mkString(", ")
+          )
+        )
+    }
+  }
+
+  // From here on, these are the shared default implementations for TypedImperativeAggregate
+
+  /** Aggregate functions which utilize HllSketch instances should never return null */
+  override def nullable: Boolean = false
+
+  /**
+   * Instantiate an HllSketch instance using the lgConfigK and tgtHllType params.
+   *
+   * @return an HllSketch instance
+   */
+  override def createAggregationBuffer(): HllSketch = {
+    new HllSketch(lgConfigK, tgtHllType)
+  }
+
+  /**
+   * Evaluate the input row and update the HllSketch instance with the row's value.
+   * The update function only supports a subset of Spark SQL types, and an
+   * UnsupportedOperationException will be thrown for unsupported types.
+   *
+   * @param sketch The HllSketch instance.
+   * @param input  an input row
+   */
+  override def update(sketch: HllSketch, input: InternalRow): HllSketch = {
+    val v = first.eval(input)
+    if (v != null) {
+      first.dataType match {
+        // Update implemented for a subset of types supported by HllSketch
+        // Spark SQL doesn't have equivalent types for ByteBuffer or char[] so leave those out
+        // Leaving out support for Array types, as unique counting these aren't a common use case
+        // Leaving out support for floating point types (IE DoubleType) due to imprecision
+        // TODO: implement support for decimal/datetime/interval types
+        case IntegerType => sketch.update(v.asInstanceOf[Int])
+        case LongType => sketch.update(v.asInstanceOf[Long])
+        case StringType => sketch.update(v.asInstanceOf[UTF8String].toString)
```

Comment:
I would expect to support also floating point types and BINARY values. How come these are missing?

Reply:
@dtenedor proposed we don't support floating point types in this comment above: https://github.com/apache/spark/pull/40615#discussion_r1154843262

I think supporting BinaryType should be do-able, though it's underlying representation would be an Array[Byte], and during our last discussion I think the guidance was we didn't expect many use cases that would require cardinality estimation of array/map/complex types.

Reply:
I am okay with leaving out floating point values for now. I personally do not see a problem supporting them, but I do not have a strong opinion.

Reply:
Regarding BINARY values: I think there is a small confusion here: in Spark we have STRING, BINARY, ARRAY, STRUCT, and MAP types. STRING and BINARY are basically arrays of characters/bytes and they are variable size data types. They are not nested, and they are essentially the same (STRINGs have additional semantics as to what should be stored, like UTF8 characters and more). ARRAY, STRUCT, and MAP are nested data types that depend on other data types (the elements of the these nested structures). We cannot really fully define them without knowing the nested types. ARRAY is variable size, STRUCT can be either fixed size or variable size (depending on the type of elements in the struct) and MAP is also variable size.

All these from the point of view of types. Then we get to the representation of those types. STRING is mapped to the `UTF8String` class as the underlying "physical" type, and BINARY is mapped to an array of bytes (what you refer to as `Array[Byte]`. BINARY is not treated as an ARRAY in the Spark type system, but rather as a single non-nested type whose "physical" representation is a Scala/Java array of bytes. So in that respect it is not any different from STRING really.

Reply:
Ok, I'm happy to add support for BinaryType - should be a simple update.